warehouse/utils/wheel.py

```diff
@@ -0,0 +1,91 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+
+import packaging.utils
+
+_PLATFORMS = [
+    (re.compile(r"^win_(.*?)$"), lambda m: f"Windows {_normalize_arch(m.group(1))}"),
+    (
+        re.compile(r"^manylinux2010_(.*?)$"),
+        lambda m: f"manylinux: glibc 2.12+ {_normalize_arch(m.group(1))}",
+    ),
+    (
+        re.compile(r"^manylinux_(\d+)_(\d+)_(.*?)$"),
+        lambda m: (
+            f"manylinux: glibc "
+            f"{m.group(1)}.{m.group(2)}+ {_normalize_arch(m.group(3))}"
+        ),
+    ),
+    (
+        re.compile(r"^musllinux_(\d+)_(\d+)_(.*?)$"),
+        lambda m: (
+            f"musllinux: musl {m.group(1)}.{m.group(2)}+ {_normalize_arch(m.group(3))}"
+        ),
+    ),
+    (
+        re.compile(r"^macosx_(\d+)_(\d+)_(.*?)$"),
+        lambda m: f"macOS {m.group(1)}.{m.group(2)}+ {_normalize_arch(m.group(3))}",
+    ),
+]
+
+_ARCHS = {
+    "amd64": "x86-64",
+    "x86_64": "x86-64",
+    "universal2": "universal2 (ARM64, x86-64)",
+    "arm64": "ARM64",
+    "armv7l": "ARMv7l",
+}
+
+
+def _normalize_arch(a: str) -> str:
+    return _ARCHS.get(a, a)
+
+
+def _format_version(s: str) -> str:
+    return f"{s[0]}.{s[1:]}"
+
+
+def filename_to_pretty_tags(filename: str) -> set[str]:
+    if not filename.endswith(".whl"):
+        return {"Source"}
+
+    try:
+        _, _, _, tags = packaging.utils.parse_wheel_filename(filename)
+    except packaging.utils.InvalidWheelFilename:
+        return set()
+
+    pretty_tags = set()
+    for tag in tags:
+        if tag.platform != "any":
+            for prefix_re, tmpl in _PLATFORMS:
+                if match := prefix_re.match(tag.platform):
+                    pretty_tags.add(tmpl(match))
+
+        if tag.interpreter.startswith("pp"):
+            # PyPy tags are a disaster, give up.
+            pretty_tags.add("PyPy")
+
+        if tag.interpreter == "py3":
+            pretty_tags.add("Python 3")
+        elif tag.interpreter == "py2":
+            pretty_tags.add("Python 2")
+        elif tag.abi == "abi3":
+            assert tag.interpreter.startswith("cp")
+            version = _format_version(tag.interpreter.removeprefix("cp"))
+            pretty_tags.add(f"CPython {version}+")
+        elif tag.abi.startswith("cp"):
+            version = _format_version(tag.abi.removeprefix("cp"))
+            pretty_tags.add(f"CPython {version}")
```

Comment:
Would it be useful to have some kind of an else-branch here for "unknown runtime"?

Reply:
Are there any cases that this currently misses? I'm not sure an unknown is actually very useful to users, simply having a missing tag may be equally valuable -- the main things these enable is people to _not_ select the wrong tag.

Reply:
My first instinct was to suggest having `typing.assert_never()` here. Now I see that it's probably not what's needed here. I wasn't sure so I figured I'd just ask. OTOH, maybe some debug logging for would be warranted this case.
Anyway, you're right — an unknown tag would probably be confusing to people.

Reply:
I think it's a good idea to at least notify us if this were to happen, so we know to update this accordingly. We can do something like this, which raises a message in Sentry with the relevant details:

https://github.com/pypi/warehouse/blob/f5b1ef3e84a52da0efc78914b1250ecddf623dc4/warehouse/oidc/models/_core.py#L179-L185

where `unaccounted_claims` would become the tag we can't handle.

Reply:
Good call. Done.

Reply:
Got the first hit on this already: `pp39-pypy39_pp73-manylinux_2_17_aarch64`. Looks like there are some filenames in the dataset I included in https://github.com/pypi/warehouse/issues/10308#issuecomment-1997665490 that aren't covered here.

Reply:
Yeah, I used that set to ensure it didn't crash, I didn't go back to verify that it never hit this sentry block.

Reply:
Also `cp36-none-win_amd64`.